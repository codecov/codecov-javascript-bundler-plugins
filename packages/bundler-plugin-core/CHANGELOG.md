# @codecov/bundler-plugin-core

## 0.0.1-beta.0

### Patch Changes

- f5cae1c: Move packages into beta

## 0.0.1-beta.0

### Patch Changes

- becd728: Add duration to webpack bundler plugin directly
- 543a526: Update dependencies before moving package out of alpha
- 4f1183e: Remove references to Webpack 4, as we currently on support Node 18+.
- c1fdbd6: Fix retry when fetching by throwing custom error if response is not okay
- d28a2e4: Break from retry loop when the response is ok
- 7c55993: Rename uploaderOverrides to uploadOverrides, remove url from uploadOverrides, and removal of repoToken/globalUploadToken in favour of just uploadToken.
- 36ed299: Alpha release of Codecov bundler plugins
- 562ac0c: Adjust peer dep versions to use x-range instead of caret
- 48b6e90: Move bundler-core test deps to dev deps
- c755a5c: Add missing outputPath to output stats file

## 0.0.1-alpha.3

### Patch Changes

- 543a526: Update dependencies before moving package out of alpha
- c1fdbd6: Fix retry when fetching by throwing custom error if response is not okay
- d28a2e4: Break from retry loop when the response is ok
- 562ac0c: Adjust peer dep versions to use x-range instead of caret
- c755a5c: Add missing outputPath to output stats file

## 0.0.1-alpha.2

### Patch Changes

- becd728: Add duration to webpack bundler plugin directly
- 7c55993: Rename uploaderOverrides to uploadOverrides, remove url from uploadOverrides, and removal of repoToken/globalUploadToken in favour of just uploadToken.

## 0.0.1-alpha.1

### Patch Changes

- 4f1183e: Remove references to Webpack 4, as we currently on support Node 18+.
- 48b6e90: Move bundler-core test deps to dev deps

## 0.0.1-alpha.0

### Patch Changes

- 36ed299: Alpha release of Codecov bundler plugins
