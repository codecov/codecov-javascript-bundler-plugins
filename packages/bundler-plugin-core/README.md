<p align="center">
  <a href="https://about.codecov.io" target="_blank">
    <img src="https://about.codecov.io/wp-content/themes/codecov/assets/brand/sentry-cobranding/logos/codecov-by-sentry-logo.svg" alt="Codecov by Sentry logo" width="280" height="84">
  </a>
</p>

# Codecov Bundler Plugin Core

> [!WARNING]  
> These plugins are currently in beta and are subject to change.

> The bundler-plugin-core is not to be used individually and is meant to be used in conjunction with the bundler specific plugins.

Core package containing the bundler-agnostic logic for the Codecov Bundler Plugins.

Checkout the individual packages for more information and examples:

- [Rollup](https://npmjs.org/package/@codecov/rollup-plugin)
- [Vite](https://npmjs.org/package/@codecov/vite-plugin)
- [Webpack](https://npmjs.org/package/@codecov/webpack-plugin)

## Features

The Codecov Bundler Core package contains the following functionality:

- Bundle Analysis Stats Upload

## More information

- [Bundler Config Docs](https://codecov.github.io/codecov-javascript-bundler-plugins/modules/_codecov_bundler_plugin_core.html)
- [Codecov Documentation](https://docs.codecov.com/docs)
- [Codecov Feedback](https://github.com/codecov/feedback/discussions)
- [Sentry Discord](https://discord.gg/Ww9hbqr)
