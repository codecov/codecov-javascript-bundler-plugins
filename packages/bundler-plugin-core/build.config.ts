import { defineBuildConfig } from "unbuild";
import { codecovRollupPlugin } from "codecovProdRollupPlugin";
import packageJson from "./package.json";

export default defineBuildConfig({
  entries: ["./src/index"],
  outDir: "dist",
  declaration: "compatible",
  sourcemap: true,
  rollup: {
    dts: {
      compilerOptions: {
        removeComments: false,
      },
    },
    emitCJS: true,
    esbuild: {
      minify: true,
    },
  },
  hooks: {
    "rollup:options": (_ctx, opts) => {
      if (process.env.PLUGIN_CODECOV_TOKEN && Array.isArray(opts.plugins)) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        opts.plugins = [
          ...opts.plugins,
          codecovRollupPlugin({
            enableBundleAnalysis:
              typeof process.env.PLUGIN_CODECOV_TOKEN === "string",
            bundleName: packageJson.name,
            uploadToken: process.env.PLUGIN_CODECOV_TOKEN,
            apiUrl: process.env.PLUGIN_CODECOV_API_URL,
          }),
        ];
      }
    },
  },
});
