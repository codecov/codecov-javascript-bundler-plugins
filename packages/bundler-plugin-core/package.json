{
  "name": "@codecov/bundler-plugin-core",
  "version": "0.0.1-beta.3",
  "description": "Official Codecov Bundler Plugin Core",
  "author": "Codecov",
  "license": "MIT",
  "type": "module",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.mts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "type-check": "tsc --noEmit",
    "build": "unbuild",
    "dev": "unbuild --stub && node --watch-path=src dist/index.mjs",
    "clean": "rm -rf dist node_modules",
    "lint": "eslint . --ext .ts,.tsx",
    "lint:fix": "pnpm lint --fix",
    "format": "prettier '**/*.{cjs,mjs,ts,tsx,md,json}' --ignore-path ../.gitignore --ignore-unknown --no-error-on-unmatched-pattern --write",
    "format:check": "prettier '**/*.{cjs,mjs,ts,tsx,md,json}' --ignore-path ../.gitignore --ignore-unknown --no-error-on-unmatched-pattern --check",
    "test:unit": "jest",
    "test:unit:ci": "jest --coverage"
  },
  "dependencies": {
    "chalk": "4.1.2",
    "semver": "^7.5.4",
    "unplugin": "^1.6.0",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@swc/core": "^1.3.107",
    "@swc/jest": "^0.2.33",
    "@types/jest": "^29.5.11",
    "@types/node": "^20.11.15",
    "@types/semver": "^7.5.6",
    "jest": "^29.7.0",
    "msw": "^2.1.5",
    "codecovProdRollupPlugin": "npm:@codecov/rollup-plugin@0.0.1-beta.2",
    "testdouble": "^3.20.1",
    "testdouble-jest": "^2.0.0",
    "ts-node": "^10.9.2",
    "typescript": "^5.3.3",
    "unbuild": "^2.0.0"
  },
  "volta": {
    "extends": "../../package.json"
  },
  "engines": {
    "node": ">=18.0.0"
  }
}
