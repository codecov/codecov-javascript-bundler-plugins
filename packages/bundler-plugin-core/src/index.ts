import { type UnpluginOptions, createUnplugin } from "unplugin";
import { satisfies } from "semver";
import { z } from "zod";

import {
  type BundleAnalysisUploadPlugin,
  type Asset,
  type Chunk,
  type Module,
  type Options,
  type Output,
} from "./types.ts";

import { jsonSchema } from "./schemas.ts";
import { red } from "./utils/logging.ts";

const NODE_VERSION_RANGE = ">=18.18.0";

interface CodecovUnpluginFactoryOptions {
  bundleAnalysisUploadPlugin: BundleAnalysisUploadPlugin;
}

export function codecovUnpluginFactory({
  bundleAnalysisUploadPlugin,
}: CodecovUnpluginFactoryOptions) {
  return createUnplugin<Options, true>((userOptions, unpluginMetaContext) => {
    const plugins: UnpluginOptions[] = [];

    if (!satisfies(process.version, NODE_VERSION_RANGE)) {
      red(
        `Codecov ${unpluginMetaContext.framework} bundler plugin requires Node.js ${NODE_VERSION_RANGE}. You are using Node.js ${process.version}. Please upgrade your Node.js version.`,
      );
      process.exit(1);
    }

    if (userOptions?.enableBundleAnalysis) {
      const statsFileName = z
        .string()
        .endsWith(".json")
        .optional()
        .parse(userOptions?.statsFileName);

      const output: Output = {
        version: "1",
      };

      let startTime = NaN;
      const { pluginVersion, version, ...pluginOpts } =
        bundleAnalysisUploadPlugin({
          output,
          statsFileName,
        });

      plugins.push({
        ...pluginOpts,
        buildStart() {
          startTime = Date.now();
          output.version = version;
          output.plugin = {
            name: pluginOpts.name,
            version: pluginVersion,
          };
          output.builtAt = startTime;
        },
        buildEnd(this) {
          const duration = Date.now() - startTime;
          output.duration = duration;
        },
      });
    }

    return plugins;
  });
}

export type { BundleAnalysisUploadPlugin, Asset, Chunk, Module, Options };
export { jsonSchema };
