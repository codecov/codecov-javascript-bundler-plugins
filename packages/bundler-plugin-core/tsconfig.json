{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {
      "@/*": ["./src/*"],
      "@test-utils/*": ["./test-utils/*"]
    },
    "checkJs": true
  },
  "include": [
    "src",
    "test-utils",
    "build.config.ts",
    "../../reset.d.ts",
    "jest.config.ts"
  ]
}
