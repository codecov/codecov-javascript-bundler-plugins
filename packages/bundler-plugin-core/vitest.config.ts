import { defineConfig } from "vitest/config";
import { config } from "../../vitest.shared";

export default defineConfig({
  ...config,
});
