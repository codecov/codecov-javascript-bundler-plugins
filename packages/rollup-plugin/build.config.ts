import { defineBuildConfig } from "unbuild";

export default defineBuildConfig({
  entries: ["./src/index"],
  outDir: "dist",
  declaration: "compatible",
  sourcemap: true,
  externals: ["rollup"],
  rollup: {
    dts: {
      compilerOptions: {
        removeComments: false,
      },
    },
    emitCJS: true,
    esbuild: {
      minify: true,
    },
  },
});
