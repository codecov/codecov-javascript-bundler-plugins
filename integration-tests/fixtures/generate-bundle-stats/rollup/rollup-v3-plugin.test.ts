/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import path from "path";
import fs from "fs";
import { type Output } from "@codecov/bundler-plugin-core";
import { rollup as rollupV3 } from "rollupV3";
// @ts-expect-error - no types
import resolve from "@rollup/plugin-node-resolve";
import commonjs from "@rollup/plugin-commonjs";
import { codecovRollupPlugin } from "@codecov/rollup-plugin";

describe("Generating rollup stats", () => {
  describe("version 3", () => {
    let stats: Output;
    const rollupPath = path.resolve(__dirname, "../../../test-apps/rollup");
    beforeAll(async () => {
      await rollupV3({
        input: `${rollupPath}/src/main.js`,
        plugins: [
          // eslint-disable-next-line @typescript-eslint/no-unsafe-call
          resolve(),
          commonjs(),
          codecovRollupPlugin({
            enableBundleAnalysis: true,
            dryRun: true,
            bundleName: "rollup-test",
          }),
        ],
      }).then((bundle) =>
        bundle.write({
          dir: `${rollupPath}/distV3`,
          entryFileNames: "[name]-[hash].js",
        }),
      );
    });

    afterAll(() => {
      fs.rm(
        path.resolve(rollupPath, "distV3"),
        { recursive: true, force: true },
        () => null,
      );
    });

    it("matches the snapshot", () => {
      const statsFilePath = path.resolve(
        rollupPath,
        "distV3/codecov-bundle-stats.json",
      );

      const statsData = fs.readFileSync(statsFilePath);
      stats = JSON.parse(statsData.toString()) as Output;

      expect(stats).toMatchSnapshot({
        builtAt: expect.any(Number),
        duration: expect.any(Number),
        outputPath: expect.stringContaining("/distV3"),
      });
    });
  });
});
