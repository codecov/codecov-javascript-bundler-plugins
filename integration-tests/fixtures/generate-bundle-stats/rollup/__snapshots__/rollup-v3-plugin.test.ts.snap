// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generating rollup stats version 3 matches the snapshot 1`] = `
{
  "assets": [
    {
      "name": "main-3e7d1fad.js",
      "normalized": "main-*.js",
      "size": 544589,
    },
  ],
  "builtAt": Any<Number>,
  "bundleName": "rollup-test-es",
  "bundler": {
    "name": "rollup",
    "version": "3.29.4",
  },
  "chunks": [
    {
      "entry": true,
      "files": [
        "main-3e7d1fad.js",
      ],
      "id": "main",
      "initial": false,
      "names": [
        "main",
      ],
      "uniqueId": "0-main",
    },
  ],
  "duration": Any<Number>,
  "modules": [
    {
      "chunkUniqueIds": [
        "0-main",
      ],
      "name": "./commonjsHelpers.js",
      "size": 315,
    },
    {
      "chunkUniqueIds": [
        "0-main",
      ],
      "name": "../node_modules/.pnpm/lodash@4.17.21/node_modules/lodash/lodash.js",
      "size": 544001,
    },
    {
      "chunkUniqueIds": [
        "0-main",
      ],
      "name": "./test-apps/rollup/src/getRandomNumber.js",
      "size": 98,
    },
    {
      "chunkUniqueIds": [
        "0-main",
      ],
      "name": "./test-apps/rollup/src/main.js",
      "size": 168,
    },
  ],
  "plugin": {
    "name": "codecov-rollup-bundle-analysis-plugin",
    "version": "1.0.0",
  },
  "version": "1",
}
`;
